### PR TITLE
Balloc tunable Framework

### PR DESCRIPTION
**Pull request by nihar1313**
Following changes contain:
->Balloc Extent db indexing.
->Framework for tuning Group Size. 
->Framework for tuning Index count.
-> Balloc code re-arrangement.

### CODE REVIEW

File: balloc/balloc.h [hunk 1]
```diff
@@ -213,6 +214,26 @@ enum {
...
+	BALLOC_DEF_BLOCK_SHIFT		= 12,
+	/**
+	 * Approximate number of groups. Exact number of groups can differ
+	 * depending on group size restrictions.
+	 */
+	BALLOC_DEF_GROUPS_NR            = 64,
+	/** Default db index count */
+	BALLOC_DEF_INDEXES_NR           = 1,
+	/** Used as minimal group size */
+	BALLOC_DEF_BLOCKS_PER_GROUP     = 524288,
```
Comment (mukundkanekar) [inline balloc/balloc.h:234]: Looks like old values of BALLOC_DEF_BLOCKS_PER_GROUP was 32768, with this new value getting Integer overflow warning during compilation.
```
balloc/ut/balloc.c:46:78: error: integer overflow in expression [-Werror=overflow]
 #define GROUP_SIZE (BALLOC_DEF_CONTAINER_SIZE / (BALLOC_DEF_BLOCKS_PER_GROUP
```

File: balloc/balloc.c [hunk 2]
```diff
@@ -156,6 +156,11 @@ static bool is_spare(uint64_t alloc_flags);
 static bool is_normal(uint64_t alloc_flags);
 static bool is_any(uint64_t alloc_flag);
 
+static int index_hash(struct m0_balloc_super_block *sb,
+		      m0_bcount_t                   block_offset)
```
Comment (mukundkanekar) [inline balloc/balloc.c:160]: To be more specific `m0_bcount_t  block_offset` can be replaced with `m0_bindex_t   ext_end`

File: balloc/balloc.c [hunk 3]
```diff
@@ -394,6 +399,9 @@ static int balloc_group_info_init(struct m0_balloc_group_info *gi,
 				 normal_zone_size,
 				 spare_zone_size, 0, 0, 0);
 #endif
+		M0_LOG(M0_INFO, "cb_db_group_desc loading grpNo : %llu  freeblk : %llu",
```
Comment (mukundkanekar) [inline balloc/balloc.c:402]: Looks like column width 80 is exceeded in above line.
"The preferred limit on the length of a single line is 80 columns"

File: balloc/balloc.c [hunk 4]
```diff
@@ -966,29 +987,26 @@ static int balloc_init_internal(struct m0_balloc *bal,
...
 		rc = -EINVAL;
 		goto out;
 	}
 
-	M0_LOG(M0_INFO, "Group Count = %"PRIu64, bal->cb_sb.bsb_groupcount);
+ 	M0_LOG(M0_INFO, "groupcount: %llu indexcount: %d blocks_per_group: %d",
+	       (unsigned long long)bal->cb_sb.bsb_groupcount,
+	       (int)bal->cb_sb.bsb_indexcount, (int)req->bfr_groupsize);
+	for (i = 0; i < bal->cb_sb.bsb_indexcount; i++) {
+		m0_be_btree_init(&bal->cb_db_group_extents[i], seg, &ge_btree_ops);
```
Comment (mukundkanekar) [inline balloc/balloc.c:1008]: looks like 80 column width is exceeded in above line.

File: balloc/balloc.c [hunk 5]
```diff
@@ -966,29 +987,26 @@ static int balloc_init_internal(struct m0_balloc *bal,
...
 	}
 
-	if (bal->cb_sb.bsb_blocksize != 1 << bshift) {
+	if (bal->cb_sb.bsb_blocksize != 1 << req->bfr_bshift) {
 		rc = -EINVAL;
 		goto out;
 	}
 
-	M0_LOG(M0_INFO, "Group Count = %"PRIu64, bal->cb_sb.bsb_groupcount);
+ 	M0_LOG(M0_INFO, "groupcount: %llu indexcount: %d blocks_per_group: %d",
```
Comment (mukundkanekar) [inline balloc/balloc.c:1004]: please remove additional space at start of line.

File: balloc/balloc.c (hunk 2 above)
Comment (nihar1313) [reply balloc/balloc.c]: done..

File: balloc/balloc.h [hunk 6]
```diff
@@ -213,6 +214,26 @@ enum {
 	M0_BALLOC_BUDDY_LOOKUP_MAX = 10,
 };
 
+/*
+ * BALLOC_DEF_BLOCKS_PER_GROUP * (1 << BALLOC_DEF_BLOCK_SHIFT) = 128 MB -->
+ * which equals group size in ext4
+ */
+enum {
```
Comment (mukundkanekar) [inline balloc/balloc.h:221]: Any reason for moving this definition up.

File: balloc/balloc.c [hunk 7]
```diff
@@ -156,6 +156,11 @@ static bool is_spare(uint64_t alloc_flags);
 static bool is_normal(uint64_t alloc_flags);
 static bool is_any(uint64_t alloc_flag);
 
+static int index_hash(struct m0_balloc_super_block *sb,
+		      m0_bcount_t                   ext_end)
```
Comment (mukundkanekar) [inline balloc/balloc.c:160]: `m0_bcount_t `and `m0_bindex_t `both are `uint64_t`.
But still use of `m0_bindex_t  `data type would be more appropriate as `ext_end `refers block index, not block count.

File: stob/ad.c [hunk 8]
```diff
@@ -285,14 +333,25 @@ static int stob_ad_domain_cfg_create_parse(const char *str_cfg_create,
...
+ 		 */ 
+		cfg->adg_group_nr = cfg->adg_container_size / (1 << cfg->adg_bshift) /
+				grp_blocks;
+		if (cfg->adg_index_nr < 1)
+			cfg->adg_index_nr = BALLOC_DEF_INDEXES_NR;
 		cfg->adg_spare_blocks_per_group =
 			m0_stob_ad_spares_calc(grp_blocks);
 		M0_LOG(M0_DEBUG, "device size %"PRId64, cfg->adg_container_size);
+		M0_LOG(M0_INFO, "device has %"PRId64" of group size %"PRId64,
+			cfg->adg_group_nr, cfg->adg_blocks_per_group);
```
Comment (mukundkanekar) [inline stob/ad.c:354]: device has total %"PRId64" groups, each with size %"PRId64

File: balloc/balloc.c [hunk 9]
```diff
@@ -861,9 +884,9 @@ static int balloc_groups_write(struct m0_balloc *bal)
 	  by this parameter.
    @return 0 means success. Otherwise, error number will be returned.
  */
-static int balloc_format(struct m0_balloc *bal,
-			 struct m0_balloc_format_req *req,
-			 struct m0_sm_group *grp)
+static int balloc_format(struct m0_balloc               *bal,
+			 struct m0_ad_balloc_format_req *req,
+			 struct m0_sm_group             *grp)
```
Comment (huanghua78) [inline balloc/balloc.c:889]: I gave different option here.
"struct m0_balloc_format_req *req" is changed to "struct m0_ad_balloc_format_req *req".
But, the "struct m0_ad_balloc_format_req" is AD specific.

"struct m0_balloc_format_req" is more general and can be used by Linux stob too.
Maybe we can still use the same struct, but add some more fields to "struct m0_balloc_format_req"?

File: motr/setup.c [hunk 10]
```diff
@@ -2231,6 +2240,16 @@ static int _args_parse(struct m0_motr *cctx, int argc, char **argv)
...
+			M0_NUMBERARG('O', "balloc no of groups per ad stob",
+				LAMBDA(void, (int64_t nr)
+				{
+					rctx->rc_balloc_group_nr = nr;
+				})),
+			M0_NUMBERARG('X', "BE balloc no of index in balloc db",
+				LAMBDA(void, (int64_t nr)
+				{
+					rctx->rc_balloc_index_nr = nr;
+				})),
```
Comment (huanghua78) [inline motr/setup.c:2252]: These parameters are only valid for mkfs, right?
For a m0d, after mkfs is done, these parameters should already be configured and cannot be changed. Is this right?

File: scripts/install/etc/sysconfig/motr [hunk 11]
```diff
@@ -94,6 +94,14 @@ MOTR_M0D_BESEG_SIZE=8589934592
 #MOTR_M0D_BETX_REG_SIZE_MAX=46137344
 #MOTR_M0D_BETX_PAYLOAD_SIZE_MAX=2097152
 
+#BALLOC configuration parameters.
+#
+#BALLOC GROUP COUNT. default is 64
+MOTR_M0D_BALLOC_GROUP_COUNT=64
+
+#BALLOC No of indexes per device, default 1
+MOTR_M0D_BALLOC_INDEX_COUNT=1
```
Comment (huanghua78) [inline scripts/install/etc/sysconfig/motr:103]: As I commented above, these two parameter can not be changed during the life time of a CORTX deployment?

File: balloc/balloc.c (hunk 9 above)
Comment (nihar1313) [reply balloc/balloc.c:889]: Had earlier tried around this, Faced the compilation header circular dependency issue.  To achieve this, we need to move m0_ad_balloc_format_req ad  as m0_balloc_format_req in balloc.h. then faced  header dependency issue.  Then latter drops that thought process as I felt balloc is more specific to AD.
Please correct me here, i may wrong here.  For Linux Stob, balloc has no role. So balloc_format_req is more specific to AD, on this thought process I used it only for AD.

File: stob/ad.c (hunk 8 above)
Comment (nihar1313) [reply stob/ad.c:354]: Done

File: balloc/balloc.c (hunk 7 above)
Comment (nihar1313) [reply balloc/balloc.c:160]: Done.

File: balloc/balloc.h (hunk 6 above)
Comment (nihar1313) [reply balloc/balloc.h:221]: Earlier Shashank also pointed this.  I am not sure, why it is showing in diff. actually there is no movement.

File: motr/setup.c (hunk 10 above)
Comment (nihar1313) [reply motr/setup.c:2252]: Yes this is correct.  Mkfs use this command line parameter. m0d always read from seg0. There are two methods
 balloc_cfg_get_from_config() from configuration parameter used by mkfs.
  balloc_cfg_get_from_adom() from m0_stob_ad_domain used by m0d.

File: scripts/install/etc/sysconfig/motr (hunk 11 above)
Comment (nihar1313) [reply scripts/install/etc/sysconfig/motr:103]: After Mkfs, it has no effect. The changed value will reflect after next time mkfs

File: balloc/balloc.c (hunk 9 above)
Comment (huanghua78) [reply balloc/balloc.c:889]: OK.

File: motr/setup.c (hunk 10 above)
Comment (huanghua78) [reply motr/setup.c:2252]: OK

File: scripts/install/etc/sysconfig/motr (hunk 11 above)
Comment (huanghua78) [reply scripts/install/etc/sysconfig/motr:103]: OK

**Review by huanghua78 — APPROVED** [review]

**Review by mukundkanekar — APPROVED** [review]

**Review by vidyadhar-pinglikar — APPROVED** [review]

**Pull request merged**